packager/certcache/certcache.go

```diff
@@ -233,20 +319,63 @@ func (this *CertCache) isHealthy(ocspResp []byte) error {
 }
 
 // Returns the OCSP response and expiry, refreshing if necessary.
-func (this *CertCache) readOCSP() ([]byte, time.Time, error) {
+// TODO(banaag): Per twifkak's suggestion, consider:
+// It may also be interesting to try to separate the retry logic from the fetch logic. One approach comes to mind:
+//
+// retryWithBackoff(func() {
+//   ocsp, err := ...
+//   return true if successful
+// }, initialWaitTime, maxTries)
+func (this *CertCache) readOCSP(allowRetries bool) ([]byte, time.Time, error) {
 	var ocspUpdateAfter time.Time
-	ocsp, err := this.ocspFile.Read(context.Background(), this.shouldUpdateOCSP, func(orig []byte) []byte {
-		return this.fetchOCSP(orig, &ocspUpdateAfter)
-	})
-	if err != nil {
-		return nil, time.Time{}, errors.Wrap(err, "Updating OCSP cache")
-	}
-	if len(ocsp) == 0 {
-		return nil, time.Time{}, errors.New("Missing OCSP response.")
+	var err error
+	var maxTries int
+
+	ocsp := []byte(nil)
+	waitTimeInMinutes := 1
+	if !allowRetries || this.certFetcher == nil {
+		// If certFetcher is nil, that means we are not auto-renewing so don't retry OCSP.
+		maxTries = 1
+	} else {
+		maxTries = maxOCSPTries
 	}
-	err = this.isHealthy(ocsp)
-	if err != nil {
-		return nil, time.Time{}, errors.Wrap(err, "OCSP failed health check.")
+
+	for numTries := 0; numTries < maxTries; {
+		this.certsMu.RLock()
+		defer this.certsMu.RUnlock()
+		ocsp, err = this.ocspFile.Read(context.Background(), this.shouldUpdateOCSP, func(orig []byte) []byte {
+			return this.fetchOCSP(orig, this.certs, &ocspUpdateAfter, numTries > 0)
+		})
+		if err != nil {
+			if numTries >= maxTries-1 {
+				return nil, time.Time{}, errors.Wrap(err, "Updating OCSP cache")
+			} else {
+				numTries++
+				waitTimeInMinutes = waitForSpecifiedTime(waitTimeInMinutes, numTries)
+				continue
+			}
+		}
+		if len(ocsp) == 0 {
```

Comment:
You have 3 places in this method that catch different error cases and trigger the numTries++;continue logic.

Consider extracting a method that does the ocsp read, checks all of the conditions (read error, empty file, unhealthy ocsp) and returns an error on any. This method and for loop then just implements the retry on error logic, but only needs a single condition/continue block.

Reply:
Added TODO.

Reply:
Done.